### PR TITLE
Add parquet parser test case for array-of-nested-structs

### PR DESCRIPTION
**Pull request by sergey-melnychuk**

### CODE REVIEW

**Comment by findepi** [comment]
do we also have product test coverage for array of struct?

**Comment by sergey-melnychuk** [comment]
> do we also have product test coverage for array of struct?

Maybe I am looking for the tests wrong, but I can't find any, at least any that specifically target this.

**Comment by sergey-melnychuk** [comment]
The added test fails due to null-values handling (an empty list is read when `null` is expected). Am I testing it wrong or it did capture a tricky edge-case?

**Comment by findepi** [comment]
> Maybe I am looking for the tests wrong, but I can't find any, at least any that specifically target this.

Let's make sure such product tests does exist.
File format are important to test on unit level, but also on integration level, covering Trino/Hive/Spark and across supported versions.

**Comment by findepi** [comment]
> Maybe I am looking for the tests wrong, but I can't find any, at least any that specifically target this.

Let's make sure such product tests does exist.
File format are important to test on unit level, but also on integration level, covering Trino/Hive/Spark and across supported versions.

Having such a test would also help answer this question:

> Am I testing it wrong or it did capture a tricky edge-case?

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestParquetArrayOfStructsColumn.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+                  optional binary value (STRING); }}}}}
+         */
+        onTrino().executeQuery("CREATE TABLE events (\n" +
+                "entries ARRAY(ROW(name VARCHAR, before ROW(value VARCHAR), after ROW(value VARCHAR))))\n" +
+                "WITH (\n" +
+                "external_location = 'hdfs:///user/hive/warehouse/TestParquetArrayOfStructsRow/events',\n" +
+                "format = 'PARQUET');");
+
+        assertThat(onTrino().executeQuery("SELECT entries FROM events"))
+                .containsExactlyInOrder(row(
```
Comment (sergey-melnychuk) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestParquetArrayOfStructsColumn.java:79]: I think I have tried all possible ways to complete the assertion but it still fails:

```
tests               | 0 - expected: 
[{name=test-entry-0, before={value=before-0}, after={value=after-0}}, {name=test-entry-1, before={value=before-1}, after={value=after-1}}, {name=test-entry-2, before={value=before-2}, after={value=after-2}}]|
tests               | 0 - actual:   
[{name=test-entry-0, before={value=before-0}, after={value=after-0}}, {name=test-entry-1, before={value=before-1}, after={value=after-1}}, {name=test-entry-2, before={value=before-2}, after={value=after-2}}]|
```

Comment (sergey-melnychuk) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestParquetArrayOfStructsColumn.java]: Leaving an array size assertion and a TODO.
Seems like implementing support of nested rows is required on Tempto side.

**Comment by sergey-melnychuk** [comment]
@findepi @sopel39 PTAL

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/parquet/AbstractTestParquetReader.java [hunk 2]
```diff
@@ -221,6 +221,60 @@ public void testCustomSchemaArrayOfStructs()
                 values, values, "self", new ArrayType(structType), Optional.of(customSchemaArrayOfStructs));
     }
 
+    @Test
+    public void testCustomSchemaArrayOfStructsIncludingStructs()
+            throws Exception
+    {
+        MessageType customSchema = parseMessageType("message CustomSchema {" +
+                "optional group self (LIST) {\n" +
```
Comment (sergey-melnychuk) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/parquet/AbstractTestParquetReader.java:229]: > The added test fails due to null-values handling (an empty list is read when `null` is expected). Am I testing it wrong or it did capture a tricky edge-case?

With `optional group self` instead of `required ...` the null-values test passes. The required field can never be null, so my intuition is that default value (empty list) was returned because of it when null-values were added explicitly by the test routine.

**Comment by sopel39** [comment]
@skrzypo987 @raunaqmorarka could you help here?

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/parquet/AbstractTestParquetReader.java (hunk 2 above)
Comment (skrzypo987) [reply plugin/trino-hive/src/test/java/io/trino/plugin/hive/parquet/AbstractTestParquetReader.java:229]: I don't think I fully understand the problem here (I don't fully understand the Parquet schema as well so don't worry).
A required field will never return null simply because it is impossible to encode it. The list will be created while reading the file without any data read.

Comment (sergey-melnychuk) [reply plugin/trino-hive/src/test/java/io/trino/plugin/hive/parquet/AbstractTestParquetReader.java:229]: This is not really a problem, just extra unit-test that checks nested arrays work properly (such test case didn't exist before or I didn't find it).

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestParquetArrayOfStructsColumn.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+              optional binary value (STRING) = 7; }}}
+         */
+        onTrino().executeQuery("CREATE TABLE events_proto (\n" +
+                "entries ARRAY(ROW(name VARCHAR, before ROW(value VARCHAR), after ROW(value VARCHAR))))\n" +
+                "WITH (\n" +
+                "external_location = 'hdfs:///user/hive/warehouse/TestParquetArrayOfStructsRow/events-proto',\n" +
+                "format = 'PARQUET');");
+
+        assertThat(onTrino().executeQuery("SELECT entries FROM events_proto"))
+                .containsExactlyInOrder(row((Object) null));
```
Comment (sergey-melnychuk) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestParquetArrayOfStructsColumn.java:114]: @skrzypo987 the problem this PR highlights is here - "proto-parquet" files are not supported (null is returned instead of content of a column - an array of nested structs). The test above checks that it works correctly with "proper parquet".

**Review by skrzypo987 — APPROVED** [review]
The test itself lgtm.

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/parquet/AbstractTestParquetReader.java [hunk 4]
```diff
@@ -226,7 +226,7 @@ public void testCustomSchemaArrayOfStructsIncludingStructs()
             throws Exception
     {
         MessageType customSchema = parseMessageType("message CustomSchema {" +
-                "required group self (LIST) {\n" +
```
Comment (sopel39) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/parquet/AbstractTestParquetReader.java]: this should be part of previous commit

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/HiveProductTest.java [hunk 5]
```diff
@@ -75,4 +81,13 @@ protected boolean isHiveVersionBefore12()
         return getHiveVersionMajor() == 0
                 || (getHiveVersionMajor() == 1 && getHiveVersionMinor() < 2);
     }
+
+    protected static void saveOnHdfs(HdfsClient hdfsClient, Path path, String location)
```
Comment (sopel39) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/HiveProductTest.java:85]: Can you inject
```
HdfsClient hdfsClient
```
within `HiveProductTest`?

nit: In follow up PR you could make `TestAvroSchemaUrl` us this new method

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestParquetArrayOfStructsColumn.java [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+{
+    @Inject
+    private HdfsClient hdfsClient;
+
+    @BeforeTestWithContext
+    public void setup()
+            throws Exception
+    {
+        hdfsClient.createDirectory("/user/hive/warehouse/TestParquetArrayOfStructsColumn/events");
+        saveOnHdfs(hdfsClient,
```
Comment (sopel39) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestParquetArrayOfStructsColumn.java:41]: nit: each arg a newline

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestParquetArrayOfStructsColumn.java [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+                Paths.get("/docker/presto-product-tests/parquet/events.parquet"),
+                "/user/hive/warehouse/TestParquetArrayOfStructsRow/events/events.parquet");
+
+        hdfsClient.createDirectory("/user/hive/warehouse/TestParquetArrayOfStructsColumn/events-proto");
+        saveOnHdfs(hdfsClient,
```
Comment (sopel39) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestParquetArrayOfStructsColumn.java:46]: nit: each arg a newline

**Review by sopel39 — APPROVED** [review]
lgtm % small comments

**Comment by bitsondatadev** [comment]
:wave: @sergey-melnychuk - this PR has become inactive. If you're still interested in working on it. It looks like this is almost there and has approvals, you just need to do a couple changes requested by @sopel39.

We're working on closing out old and inactive PRs, so if you're too busy or this has too many merge conflicts to be worth picking back up, we'll be making another pass to close it out in a few weeks.
